### PR TITLE
CFE-3991: Added version_compare() policy function

### PR DESCRIPTION
**Pull request by olehermanse**
This function can be used to compare 2 arbitrary semver version
numbers. This can be useful if you have 2 versions of a package
and you want to know if they are the same version, if one is
newer than the other etc.

Internally, it uses the same version comparison logic as
the cf_version functions (cf_version_minimum, etc.). This means
that only the major.minor.patch parts of the version string
are compared, the rest is discarded, so
3.2.1 = 3.2.1-1 = 3.2.1-2 etc.

### CODE REVIEW

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9940/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9940/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9940/

**Comment by olehermanse** [comment]
Currently failing because this one has not been merged yet: https://github.com/cfengine/core/pull/5339

**Review by nickanderson — APPROVED** [review]
Good, suggested expansion of example to show part of string that is discarded. But still ok without that.

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -2229,6 +2229,40 @@ static VersionComparison GenericVersionCheck(
...
+        return FnFailure();
+    }
+    const char *const version_a = RlistScalarValue(args);
+    const char *const operator = RlistScalarValue(args->next);
+    const char *const version_b = RlistScalarValue(args->next->next);
+
+    const BooleanOrError result = CompareVersionExpression(version_a, operator, version_b);
+    if (result == BOOLEAN_ERROR) {
+        Log(LOG_LEVEL_ERR,
+            "Cannot comapre versions: %s(\"%s\", \"%s\", \"%s\")",
```
Comment (vpodzime) [inline libpromises/evalfunction.c:2256]: *comapre!*

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -9329,6 +9363,14 @@ static const FnCallArg CFVERSIONBETWEEN_ARGS[] =
     {NULL, CF_DATA_TYPE_NONE, NULL}
 };
 
+static const FnCallArg VERSION_COMPARE_ARGS[] =
+{
+    {CF_ANYSTRING, CF_DATA_TYPE_STRING, "First version to compare"},
+    {CF_ANYSTRING, CF_DATA_TYPE_STRING, "Operator to use in comparison (=, !=, <, >, <=, >=)"},
```
Comment (vpodzime) [inline libpromises/evalfunction.c:9369]: Shouldn't this use a regex for the allowed strings?

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -2229,6 +2229,40 @@ static VersionComparison GenericVersionCheck(
...
+        ARG_UNUSED const Policy *policy,
+        const FnCall *fp,
+        const Rlist *args)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+    if (args == NULL || args->next == NULL || args->next->next == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Policy fuction %s requires 3 arguments:"
```
Comment (craigcomstock) [inline libpromises/evalfunction.c:2243]:
```suggestion
            "Policy function %s requires 3 arguments:"
```

:O

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -10570,6 +10612,8 @@ const FnCallType CF_FNCALL_TYPES[] =
                   FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
     FnCallTypeNew("variablesmatching", CF_DATA_TYPE_STRING_LIST, CLASSMATCH_ARGS, &FnCallVariablesMatching, "List the variables matching regex arg1 and tag regexes arg2,arg3,...",
                   FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("cf_version_minimum", CF_DATA_TYPE_CONTEXT, VERSION_COMPARE_ARGS, &FnCallVersionCompare, "Compare two version numbers with a specified operator (=, !=, <, >, <=, >=)",
```
Comment (craigcomstock) [inline libpromises/evalfunction.c:10615]:
```suggestion
    FnCallTypeNew("version_compare", CF_DATA_TYPE_CONTEXT, VERSION_COMPARE_ARGS, &FnCallVersionCompare, "Compare two version numbers with a specified operator (=, !=, <, >, <=, >=)",
```

**Review by craigcomstock — CHANGES_REQUESTED** [review]
a few typos I think

File: libpromises/evalfunction.c (hunk 4 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:10615]: Thanks!

File: libpromises/evalfunction.c (hunk 3 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:2243]: Thanks!

File: libpromises/evalfunction.c (hunk 1 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:2256]: Nice one, fixed :)

File: examples/version_compare.cf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#+begin_src cfengine3
+bundle agent __main__
+{
+  reports:
+    "This will be skipped because the version comparison is false"
+      if => version_compare("3.21.0", "<", "3.20.99");
+    "This will be printed because the version comparison is true"
+      if => version_compare("3.24.1", ">=", "3.24.1");
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply examples/version_compare.cf:8]: I disagree, I want the examples to be very simple. This can be covered by docs and tests.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please :)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9943/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9943/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9943/

File: libpromises/evalfunction.c [hunk 6]
```diff
@@ -10570,6 +10612,8 @@ const FnCallType CF_FNCALL_TYPES[] =
                   FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
     FnCallTypeNew("variablesmatching", CF_DATA_TYPE_STRING_LIST, CLASSMATCH_ARGS, &FnCallVariablesMatching, "List the variables matching regex arg1 and tag regexes arg2,arg3,...",
                   FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("version_compare", CF_DATA_TYPE_CONTEXT, VERSION_COMPARE_ARGS, &FnCallVersionCompare, "Compare two version numbers with a specified operator (=, !=, <, >, <=, >=)",
```
Comment (nickanderson) [inline libpromises/evalfunction.c:10615]: I see that you have `=` here, but not `==` which is used in one of the tests.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Review by nickanderson — DISMISSED** [review]

**Review by craigcomstock — DISMISSED** [review]
cool, wait for jenkins then yeah? 👍

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9945/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9945/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9945/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9946/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9946/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9946/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/9947/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/9947/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-9947/

**Review by larsewi — APPROVED** [review]
Nice 🚀

**Review by craigcomstock — APPROVED** [review]

**Pull request merged**